### PR TITLE
[EXPERIMENTAL][DI] Add tail injection for methods + autowiring

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

ping @fabpot @simensen

```yaml
services:
    foo:
        class: Foo
        tails:
            method1: { 1: 'bar', 2: '@bar' }
```

will create an inheritance proxy overriding `Foo::method1`, making its args 2 & 3 optional, with the configured values being used as defaults, keeping only argument 1 mandatory (if it was in the original declaration of course).

Allows Laravel like controllers (`function indexAction(Request $request, PostRepository $postRepository)` with `$postRepository` injection), but with full Symfony-style flexibility (ie you can/have to exactly choose where+what to inject).
Note that this can *not* be achieved via a 3.2 `ArgumentResolver` : only compile time can achieve the required level of configurability - not runtime, which `ArgumentResolver` is only about.
Works with any service really of course.
Reuses the inheritance proxy generation already in place for getter injection.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Note that autowiring will be added later - the PR is big enough.
The way I see it requires https://github.com/symfony/symfony/pull/21194
And declaration in Yaml would look like:
```yaml
services:
  App\Controller\FooController:
    autowire:
      - !tail '*Action'
```

**Comment by stof** [comment]
this requires that the arguments are at the end, right ?

**Comment by nicolas-grekas** [comment]
yes, it throws if there are remaining required tail args.

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 1]
```diff
@@ -1325,40 +1335,161 @@ private function generateOverriddenGetters($id, Definition $definition, \Reflect
...
         if ($r->isStatic()) {
-            throw new RuntimeException(sprintf('Unable to configure getter injection for service "%s": method "%s::%s" cannot be static.', $id, $class->name, $r->name));
+            throw new RuntimeException(sprintf('Unable to configure %s for service "%s": method "%s::%s" cannot be static.', $msg, $id, $class->name, $r->name));
         }
         if ($r->isFinal()) {
-            throw new RuntimeException(sprintf('Unable to configure getter injection for service "%s": method "%s::%s" cannot be marked as final.', $id, $class->name, $r->name));
+            throw new RuntimeException(sprintf('Unable to configure %s for service "%s": method "%s::%s" cannot be marked as final.', $msg, $id, $class->name, $r->name));
         }
+        if ($type = method_exists($r, 'getReturnType') ? $r->getReturnType() : null) {
+            $type = ': '.($type->allowsNull() ? '?' : '').self::generateTypeHint($type, $r);
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: I vote for moving the building of the dumped return type outside this method. I already found the reference argument harder to read previously, but it becomes worse when nesting the resolution by reference in another method. We already return all the necessary info anyway, as this logic depends only on the reflector anyway

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: done

File: src/Symfony/Component/DependencyInjection/ContainerBuilder.php [hunk 2]
```diff
@@ -1488,4 +1622,17 @@ private function getExpressionLanguage()
...
+    {
+        if (!is_array($value)) {
+            return var_export($value, true);
+        }
+        $code = array();
+        foreach ($value as $k => $v) {
+            $code[] = sprintf('%s => %s', var_export($k, true), self::export($v));
+        }
+
+        return sprintf('array(%s)', implode(', ', $code));
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: isn't it exactly what ``var_export($array)`` does ?

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/ContainerBuilder.php]: with one difference: this generates a single line export, while var_export splits arrays on several lines

**Comment by nicolas-grekas** [comment]
PR is ready, with tests and autowiring.

File: src/Symfony/Component/DependencyInjection/ChildDefinition.php [hunk 3]
```diff
@@ -162,11 +162,21 @@ public function setDeprecated($boolean = true, $template = null)
     /**
      * {@inheritdoc}
      */
-    public function setAutowiredMethods(array $autowiredMethods)
+    public function setAutowiredCalls(array $autowiredCalls)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/ChildDefinition.php]: @dunglas @stof if you're ok with this rename, I can submit a refactoring PR first, to reduce the diff of this PR, OK for you?

Comment (stof) [reply src/Symfony/Component/DependencyInjection/ChildDefinition.php]: looks good to me. And doing it now is still possible as the existing name is not yet released

**Comment by nicolas-grekas** [comment]
"partial" renamed to "tail"

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 4]
```diff
@@ -33,6 +33,7 @@ class AutowirePass extends AbstractRecursivePass
      * @internal
      */
     const MODE_OPTIONAL = 2;
+    const MODE_TAIL = 3;
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: should also be ``@internal``

**Comment by stof** [comment]
@nicolas-grekas why is this PR opened on your fork rather than on the symfony repo ? this looks weird

**Comment by nicolas-grekas** [comment]
@stof because this is going to be even more surprising than getter injection to some :)
I'd like the idea to be confirmed first with a small core-team related group.
What do you think yourself of the usefulness of this?

**Pull request closed without merging**